### ISSUE TITLE
Unable to specify "big messages" bucket for ServiceControl

### ISSUE THREAD

**Issue by mattcanty** [Particular/ServiceControl#1616]
We're seeing this error with large messages, but it appears that it is not possible to specify a bucket.

```
2019-04-01 12:45:13.0159|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 31407dd0-b5ee-4427-83c5-0cdf281dcf11 as a poison message due to exception Amazon.S3.AmazonS3Exception: The specified bucket does not exist ---> Amazon.Runtime.Internal.HttpErrorResponseException: The remote server returned an error: (404) Not Found. ---> System.Net.WebException: The remote server returned an error: (404) Not Found.
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.HttpRequest.<GetResponseAsync>d__16.MoveNext()
   --- End of inner exception stack trace ---
   at Amazon.Runtime.Internal.HttpRequest.<GetResponseAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.HttpHandler`1.<InvokeAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.RedirectHandler.<InvokeAsync>d__1`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.Unmarshaller.<InvokeAsync>d__3`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.S3.Internal.AmazonS3ResponseHandler.<InvokeAsync>d__1`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.ErrorHandler.<InvokeAsync>d__5`1.MoveNext()
   --- End of inner exception stack trace ---
   at Amazon.Runtime.Internal.HttpErrorResponseExceptionHandler.HandleException(IExecutionContext executionContext, HttpErrorResponseException exception)
   at Amazon.Runtime.Internal.ErrorHandler.ProcessException(IExecutionContext executionContext, Exception exception)
   at Amazon.Runtime.Internal.ErrorHandler.<InvokeAsync>d__5`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.CallbackHandler.<InvokeAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.CredentialsRetriever.<InvokeAsync>d__7`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.RetryHandler.<InvokeAsync>d__10`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Amazon.Runtime.Internal.RetryHandler.<InvokeAsync>d__10`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.CallbackHandler.<InvokeAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.CallbackHandler.<InvokeAsync>d__9`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.S3.Internal.AmazonS3ExceptionHandler.<InvokeAsync>d__1`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.ErrorCallbackHandler.<InvokeAsync>d__5`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Amazon.Runtime.Internal.MetricsHandler.<InvokeAsync>d__1`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.AmazonSQS.MessageExtensions.<ToIncomingMessage>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transports.SQS.MessagePump.<ProcessMessage>d__7.MoveNext(). Moving to error queue.
2019-04-01 12:45:13.0159|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 31407dd0-b5ee-4427-83c5-0cdf281dcf11 as a poison message because it could not be converted to an IncomingMessage. Moving to error queue.
```

Tried setting the following but not working: 

```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
    </connectionStrings>
        <add name="NServiceBus/Transport" connectionString="Region=eu-west-1;QueueNamePrefix=Staging-;S3BucketForLargeMessages=staging-nservicebus-bigmessages;" />
    </connectionStrings>
</configuration>
```

After overwriting the binaries in order to pass in S3BucketForLargeMessages and S3KeyPrefix to [ServiceControl.Transports.SQS](https://github.com/Particular/ServiceControl/blob/develop/src/ServiceControl.Transports.SQS/SQSTransportCustomization.cs#L31) we get the following error:

```
2019-04-01 13:49:38.6062|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 456bd3cd-3720-408a-a2de-245c21e91e7d as a poison message due to exception System.NotSupportedException: This stream does not support seek operations.
   at System.Net.ConnectStream.get_Length()
   at NServiceBus.AmazonSQS.MessageExtensions.<ToIncomingMessage>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transports.SQS.MessagePump.<ProcessMessage>d__7.MoveNext(). Moving to error queue.
2019-04-01 13:49:38.6218|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 456bd3cd-3720-408a-a2de-245c21e91e7d as a poison message because it could not be converted to an IncomingMessage. Moving to error queue.
```

We believe there is an issue here https://github.com/Particular/NServiceBus.AmazonSQS/blob/develop/src/NServiceBus.AmazonSQS/Extensions/MessageExtensions.cs#L25

**EDIT** We've identified with `MessagePump` issue and I have raised a PR with NServiceBus.AmazonSQS https://github.com/Particular/NServiceBus.AmazonSQS/pull/290

**Comment by mattcanty** [comment]
After overwriting the binaries we face this error:

```
2019-04-01 13:49:38.6062|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 456bd3cd-3720-408a-a2de-245c21e91e7d as a poison message due to exception System.NotSupportedException: This stream does not support seek operations.
   at System.Net.ConnectStream.get_Length()
   at NServiceBus.AmazonSQS.MessageExtensions.<ToIncomingMessage>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transports.SQS.MessagePump.<ProcessMessage>d__7.MoveNext(). Moving to error queue.
2019-04-01 13:49:38.6218|31|Warn|NServiceBus.Transports.SQS.MessagePump|Treating message with SQS Message Id 456bd3cd-3720-408a-a2de-245c21e91e7d as a poison message because it could not be converted to an IncomingMessage. Moving to error queue.
```

**Comment by danielmarbach** [comment]
Currently, ServiceControl does not allow to specify the bucket for large messages. It was probably an oversight of us that we didn't add it as a connection string parameter. 

So you have messages with large bodies and when you try to retry them ServiceControl will try to use the body key to retrieve the body and then fail is that correct?

I think we have to think through this scenario a bit because if we would allow specifying the bucket what would happen on retrying batches is that the staging logic would download the body payload only to then re-upload it again to storage.

**Comment by mattcanty** [comment]
Yeah when publishing a large message it is immediately put onto ServiceControl's error queue. It fails because the bucket name is defaulted to `string.empty`.

Managed to get over this then hit the message pump issue.

I get what you're saying, downloading and uploading immediately seems like a poor use of resources.

**Comment by danielmarbach** [comment]
The thing is we should not even mess with the bucket when do retry messages. The payload is already on S3 and we could just forward the message again into the input queue. But again I think I might be jumping to conclusions here and we have to have a deeper look because ServiceControl itself has the ability to store the body and thus when the messages get retried with bodies larger than 256kb new S3 objects are created. The benefit of this is that you are no longer subjected to the S3 expiry policy of the original endpoint because ingested error messages get stored according to the expiry policy of ServiceControl. 

Which environment of yours does this affect currently?

**Comment by mattcanty** [comment]
We run NServiceBus and ServiceControl in Production on MSMQ. Found this issue during migration to SQS in Production. We phased the approach so zero downtime, but ultimately we're rolled back until we can resolve.

Have licensed version of NServiceBus.

**Comment by danielmarbach** [comment]
I raised an internal issue to triage it and will update you with further details once we know more

**Comment by mattcanty** [comment]
Hi @danielmarbach has there been any progress? We're still running a patch at the moment.

**Comment by danielmarbach** [comment]
PR raised https://github.com/Particular/ServiceControl/pull/1634

**Comment by danielmarbach** [comment]
Fixed as part of 3.7.1, see https://discuss.particular.net/t/servicecontrol-3-7-1-patch-release-available/1126
